### PR TITLE
Feature/self-consistency-prompting

### PR DESCRIPTION
**Pull request by henry-ppp**
### **Summary**
This PR implements a new self-consistency prompting strategy for test case generation in the Reflexion framework. The implementation follows the established patterns in the codebase and integrates seamlessly with existing functionality.

#### **1. New Prompt Constants** (`prompt_constants.py`)
- Added `RB_SELF_CONSISTENCY_TEST_GENERATION_CHAT_INSTRUCTION`: Main instruction for self-consistency test generation
- Added `RB_SELF_CONSISTENCY_TEST_GENERATION_FEW_SHOT`: Examples demonstrating the self-consistency method with edge cases like `a = b`, `a = 1 or b = 1`, and `a much larger than b`

#### **2. New Utility Function** (`generator_utils.py`)
- **Features**:
  - Filters out inconsistent test cases (only keeps `"CONSISTENT"` ones)
  - Converts from self-consistency format to standard `(input, output)` format
  - Handles both single test cases and lists of test cases
  - Gracefully handles parsing errors
  - Follows the same interface as existing test generation utilities

#### **3. Enhanced PyGenerator** (`py_generate.py`)
- Uses the new utility function for clean separation of concerns
- Maintains consistent parameter ordering with other test generation methods

### **How It Works**

The self-consistency strategy:
1. **Generates test cases** using step-by-step reasoning prompts
2. **Validates consistency** by comparing initial guesses against reasoned results
3. **Filters results** to only return logically consistent test cases
4. **Converts format** from detailed self-consistency structure to standard test format

### **Usage Example**
```python
from generators import generator_factory, model_factory

gen = generator_factory("python")
model = model_factory("gpt-4")

# Generate self-consistency tests
test_cases = gen.self_consistency_tests(
    samples=samples,
    problem_context=problem_context,
    inferred_specificaion=inferred_specification,
    func=buggy_func,
    model=model,
    max_num_tests=5
)
```

---

**Files Changed:**
- `apr/generators/prompt_constants.py` - Added self-consistency prompt constants
- `apr/generators/generator_utils.py`
- `apr/generators/py_generate.py` - Added `self_consistency_tests()` method to PyGenerator

**No breaking changes** - all existing functionality remains intact.

### CODE REVIEW

File: apr/generators/prompt_constants.py [hunk 1]
```diff
@@ -1306,4 +1306,98 @@
...
+- The problem description,  
+- The input format,  
+- The expected output format,  
+- The pre-run execution outcome.
+
+**Role**: Software engineer. Your task is to apply the **self-consistency method** to generate and validate test cases for a correct Ruby implementation.
+
+Please organize the test cases into the following category:
+
+**1. Self-Consistency for Expected Outputs**:
```
Comment (niktaakbarpour) [inline apr/generators/prompt_constants.py]: We do not need this line at all

File: apr/generators/prompt_constants.py [hunk 2]
```diff
@@ -1306,4 +1306,98 @@
...
+RB_SELF_CONSISTENCY_TEST_GENERATION_CHAT_INSTRUCTION = """You are an AI Ruby programming language coding assistant tasked with generating high-quality test cases based on the provided problem context, which includes:  
+- The buggy source code,  
+- The problem description,  
+- The input format,  
+- The expected output format,  
+- The pre-run execution outcome.
+
+**Role**: Software engineer. Your task is to apply the **self-consistency method** to generate and validate test cases for a correct Ruby implementation.
+
+Please organize the test cases into the following category:
```
Comment (niktaakbarpour) [inline apr/generators/prompt_constants.py]: You do not say what the categories are.

File: apr/generators/prompt_constants.py [hunk 3]
```diff
@@ -1306,4 +1306,98 @@
...
+"""
+
+RB_SELF_CONSISTENCY_TEST_GENERATION_CHAT_INSTRUCTION = """You are an AI Ruby programming language coding assistant tasked with generating high-quality test cases based on the provided problem context, which includes:  
+- The buggy source code,  
+- The problem description,  
+- The input format,  
+- The expected output format,  
+- The pre-run execution outcome.
+
+**Role**: Software engineer. Your task is to apply the **self-consistency method** to generate and validate test cases for a correct Ruby implementation.
```
Comment (niktaakbarpour) [inline apr/generators/prompt_constants.py]: Remove this line. Just add what should the agent do. This line is not correct.

File: apr/generators/prompt_constants.py [hunk 4]
```diff
@@ -1306,4 +1306,98 @@
...
+- Do **NOT** include:
+  - Function implementations  
+  - Test runners or harnesses  
+  - Commentary or extra prose
+
+- Ensure:
+  - Inputs follow the problem's format and constraints  
+  - Outputs reflect the intended correct behavior  
+  - Return test cases as a valid JSON array"""
+
```
Comment (niktaakbarpour) [inline apr/generators/prompt_constants.py]: look at this prompt: RB_TEST_GENERATION_IO_CHAT_INSTRUCTION
and write something in that format but for self-consistency.

File: docs/Reflextion Essay Summary [hunk 5]
```diff
@@ -0,0 +1,135 @@
+Here is a **detailed long-form summary** of the paper *“Reflexion: Language Agents with Verbal Reinforcement Learning”* by Noah Shinn et al.:
```
Comment (niktaakbarpour) [inline docs/Reflextion Essay Summary]: Remove this file

File: apr/generators/py_generate.py [hunk 6]
```diff
@@ -168,6 +172,19 @@ def internal_tests(self, samples: List[str], problem_context: str, inferred_spec
...
         )
     
+    def self_consistency_tests(self, samples: List[str], problem_context: str, inferred_specificaion:str, func: str, model: ModelBase, max_num_tests: int = 7) -> List[str]:
+        """Generate test cases using self-consistency prompting strategy."""
+        return generic_generate_self_consistency_tests(
+            samples=samples,
+            problem_context=problem_context,
+            model=model,
+            max_num_tests=max_num_tests,
+            test_generation_few_shot=RB_SELF_CONSISTENCY_TEST_GENERATION_FEW_SHOT,
```
Comment (niktaakbarpour) [inline apr/generators/py_generate.py]: Change the names of test_generation_few_shot and test_generation_chat_instruction. Add something meaningful.

File: apr/generators/py_generate.py [hunk 7]
```diff
@@ -168,6 +172,19 @@ def internal_tests(self, samples: List[str], problem_context: str, inferred_spec
...
+            test_generation_chat_instruction=RB_SELF_CONSISTENCY_TEST_GENERATION_CHAT_INSTRUCTION,
+            test_generation_completion_instruction=PY_TEST_GENERATION_COMPLETION_INSTRUCTION,
```
Comment (niktaakbarpour) [inline apr/generators/py_generate.py]: Pay attention to the inputs. Do we really need this here?
test_generation_completion_instruction=PY_TEST_GENERATION_COMPLETION_INSTRUCTION,

File: docs/Relexion_Language Agents with Verbal Reinforcement Learning.pdf
Comment (niktaakbarpour) [inline docs/Relexion_Language Agents with Verbal Reinforcement Learning.pdf:1]: Remove this file

**Review by niktaakbarpour — CHANGES_REQUESTED** [review]

**Comment by henry-ppp** [comment]
## 🔧 **Improve Parameter Naming and Prompt for Self-Consistency Test Generation**

### **Summary**
This PR improves parameter naming and Prompt in the self-consistency test generation function for better code readability and maintainability.

### **Key Changes**

#### **1. Improved Parameter Names** (`generator_utils.py`)
- Changed `test_generation_few_shot` → `self_consistency_test_generation_few_shot`
- Changed `test_generation_chat_instruction` → `self_consistency_test_generation_chat_instruction`
- Removed unnecessary `test_generation_completion_instruction` parameter

#### **2. Updated Function Calls** (`py_generate.py`)
- Updated `self_consistency_tests()` method to use new parameter names

#### **3. Updated Prompt Constants** (`prompt_constants.py`)
- Updated references to use the new parameter names in the self-consistency function
---

**Files Changed:**
- `apr/generators/generator_utils.py`
- `apr/generators/py_generate.py`
- `apr/generators/prompt_constants.py`

**No breaking changes** - all existing functionality remains intact.
